bitvm/src/bn254/g1.rs

```diff
@@ -0,0 +1,1807 @@
+            { Fq::is_zero_keep_element(1) }
+            OP_FROMALTSTACK OP_BOOLAND
+            OP_IF // if x == 0 and y == 0, then z = 0
+                { Fq::push_zero() }
+            OP_ELSE // else z = 1
+                { Fq::push_one() }
+            OP_ENDIF
+        }
+    }
+
+    pub fn is_zero() -> Script {
+        script! {
+            { Fq::is_zero(0) }
+            OP_TOALTSTACK
+            { Fq::is_zero(0) }
+            OP_FROMALTSTACK
+            OP_BOOLAND
+        }
+    }
+
+    pub fn is_zero_keep_element() -> Script {
+        script! {
+            { Fq::is_zero_keep_element(0) }
+            OP_TOALTSTACK
+            { Fq::is_zero_keep_element(1) }
+            OP_FROMALTSTACK
+            OP_BOOLAND
+        }
+    }
+
+    pub fn drop() -> Script {
+        script! {
+            { Fq::drop() }
+            { Fq::drop() }
+        }
+    }
+
+    pub fn roll(mut a: u32) -> Script {
+        a *= 2;
+        script! {
+            { Fq::roll(a + 1) }
+            { Fq::roll(a + 1) }
+        }
+    }
+}
+
+/// input of func (params):
+///      p.x, p.y
+/// output on stack:
+///      x' = -p.x / p.y
+///      y' = 1 / p.y
+pub fn from_eval_point(p: ark_bn254::G1Affine) -> Script {
+    let py_inv = p.y().unwrap().inverse().unwrap();
+    script! {
+        { Fq::push_u32_le(&BigUint::from(py_inv).to_u32_digits()) }
+        // [1/y]
+        // check p.y.inv() is valid
+        { Fq::copy(0) }
+        // [1/y, 1/y]
+        { Fq::push_u32_le(&BigUint::from(p.y).to_u32_digits()) }
+        // [1/y, 1/y, y]
+        { Fq::mul() }
+        // [1/y, 1]
+        { Fq::push_one() }
+        // [1/y, 1, 1]
+        { Fq::equalverify(1, 0) }
+        // [1/y]
+
+        // -p.x / p.y
+        { Fq::copy(0) }
+        // [1/y, 1/y]
+        { Fq::push_u32_le(&BigUint::from(p.x).to_u32_digits()) }
+        // [1/y, 1/y, x]
+        { Fq::neg(0) }
+        // [1/y, 1/y, -x]
+        { Fq::mul() }
+        // [1/y, -x/y]
+        { Fq::roll(1) }
+        // [-x/y, 1/y]
+    }
+}
+
+
+/// input of func (params):
+///      p.x, p.y
+/// Input Hints On Stack
+///      tmul hints, p.y_inverse
+/// output on stack:
+///      x' = -p.x / p.y
+pub fn hinted_x_from_eval_point(p: ark_bn254::G1Affine, py_inv: ark_bn254::Fq) -> (Script, Vec<Hint>) {
+    let mut hints = Vec::new();
+
+    let (hinted_script1, hint1) = Fq::hinted_mul(1, p.y, 0, py_inv);
+    let (hinted_script2, hint2) = Fq::hinted_mul(1, py_inv, 0, -p.x);
+    let script_lines = vec! [
+        // Stack: [hints, pyd, px, py] 
+        Fq::copy(2),
+        // Stack: [hints, pyd, px, py, pyd] 
+        hinted_script1,
+        Fq::push_one_not_montgomery(),
+        Fq::equalverify(1, 0),
+        // Stack: [hints, pyd, px]
+        Fq::neg(0),
+        // Stack: [hints, pyd, -px]
+        hinted_script2
+    ];
+
+    let mut script = script!{};
+    for script_line in script_lines {
+        script = script.push_script(script_line.compile());
+    }
```

Comment:
Is the `script_lines` part necessary here or could we use the `script!{}` macro?